classes/condition.php

```diff
@@ -23,163 +27,135 @@
-     * @return bool True if available
-     * @throws \dml_exception
+     * @throws dml_exception
      */
-    public function is_available($not, info $info, $grabthelot, $userid) {
+    public function is_available($not, info $info, $grabthelot, $userid): bool {
         if ($grabthelot) {
             self::preload_exceptions($userid);
         }
-
         $course = $info->get_course();
         $ret = self::is_course_available($course->id, $userid);
-
         return ($not xor $ret);
     }
 
     /**
-     * Obtains a string describing this restriction (whether or not
-     * it actually applies). Used to obtain information that is displayed to
-     * students if the activity is not available to them, and for staff to see
-     * what conditions are.
-     *
-     * The $full parameter can be used to distinguish between 'staff' cases
-     * (when displaying all information about the activity) and 'student' cases
-     * (when displaying only conditions they don't meet).
-     *
-     * If implementations require a course or modinfo, they should use
-     * the get methods in $info.
-     *
-     * The special string <AVAILABILITY_CMNAME_123/> can be returned, where
-     * 123 is any number. It will be replaced with the correctly-formatted
-     * name for that activity.
+     * {@inheritDoc}
      *
      * @param bool $full Set true if this is the 'full information' view
      * @param bool $not Set true if we are inverting the condition
      * @param info $info Item we're checking
-     * @return string Information string (for admin) about all restrictions on
-     *   this item
-     * @throws \moodle_exception
+     * @throws moodle_exception
      */
-    public function get_description($full, $not, info $info) {
+    public function get_description($full, $not, info $info): string {
         global $USER;
-
         $course = $info->get_course();
-
         if ($not) {
             $str = get_string('requires_no2fa', 'availability_shibboleth2fa');
         } else {
             $str = get_string('requires_2fa', 'availability_shibboleth2fa');
-
             if (!$full || !$this->is_available($not, $info, false, $USER->id)) {
-                $url = new \moodle_url('/availability/condition/shibboleth2fa/index.php', array('id' => $course->id));
+                $url = new moodle_url('/availability/condition/shibboleth2fa/index.php', ['id' => $course->id]);
                 if ($info instanceof info_module) {
                     $url->param('cmid', $info->get_course_module()->id);
                 } else if ($info instanceof info_section) {
                     $url->param('sectionid', $info->get_section()->section);
                 }
-                $str = \html_writer::link($url, $str);
+                $str = html_writer::link($url, $str);
             }
         }
-
         if ($full) {
-            $context = \context_course::instance($course->id, MUST_EXIST);
+            $context = context_course::instance($course->id);
```

Comment:
I understand what you mean, but I have to disagree. If anything, the `instance` method should be properly documented by Moodle, instead of completely disregarding the `$strictness` parameter or its implication for the return type/value.

But regardless, I don't think there is any potential confusion here. Reading this and the following few lines makes me expect nothing other than a `context_course` type object in `$context` and some sort of error, if no such course exists. I cannot remember any time I explicitly repeated a default argument to any function to be honest. (Aside from old PHP that _forces_ you to repeat it, if you want to pass a non-default positional argument after it. But this thankfully became unnecessary with named arguments.)

And even the mantra [_explicit is better than implicit_](https://peps.python.org/pep-0020/) has its limits.